From a9f342a286225f539c313eb9f19f274f2c3c273d Mon Sep 17 00:00:00 2001
From: Andreas Larsson <andreas@gaisler.com>
Date: Tue, 15 Sep 2020 13:49:27 +0200
Subject: [PATCH 29/50] common: Add errata fix for TN-0009 and TN-0010 and
 adjust errata fix for TN-0011

This adds missing fixes for TN-0009 and TN-0010 and applies the TN-0011
fix only when needed.
---
 libstdc++-v3/config/cpu/sparc/atomicity.h | 38 ++++++++++++++++++++---
 1 file changed, 34 insertions(+), 4 deletions(-)

diff --git a/libstdc++-v3/config/cpu/sparc/atomicity.h b/libstdc++-v3/config/cpu/sparc/atomicity.h
index 368a9c28666..9f01d675b09 100644
--- a/libstdc++-v3/config/cpu/sparc/atomicity.h
+++ b/libstdc++-v3/config/cpu/sparc/atomicity.h
@@ -82,7 +82,13 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
   __exchange_and_add(volatile _Atomic_word* __mem, int __val) throw ()
   {
     _Atomic_word __result, __tmp;
-    __asm__ __volatile__(".align 16\n\t"
+    __asm__ __volatile__(
+#ifdef __FIX_LEON3FT_TN0010
+			 "nop\n\t"
+#endif
+#ifdef __FIX_LEON3FT_TN0011
+			 ".align 16\n\t"
+#endif
 			 "1:	ldstub	[%1], %0\n\t"
 			 "	cmp	%0, 0\n\t"
 			 "	bne	1b\n\t"
@@ -92,7 +98,16 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 			 : "memory");
     __result = *__mem;
     *__mem += __val;
-    __asm__ __volatile__("stb	%%g0, [%0]"
+    __asm__ __volatile__(
+#ifdef __FIX_LEON3FT_B2BST
+			 "nop\n\t"
+			 "nop\n\t"
+#endif
+			 "stb	%%g0, [%0]\n\t"
+#ifdef __FIX_LEON3FT_B2BST
+			 "nop\n\t"
+			 "nop\n\t"
+#endif
 			 : /* no outputs */
 			 : "r" (&_Atomicity_lock<0>::_S_atomicity_lock)
 			 : "memory");
@@ -104,7 +119,13 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
   __atomic_add(volatile _Atomic_word* __mem, int __val) throw ()
   {
     _Atomic_word __tmp;
-    __asm__ __volatile__(".align 16\n\t"
+    __asm__ __volatile__(
+#ifdef __FIX_LEON3FT_TN0010
+			 "nop\n\t"
+#endif
+#ifdef __FIX_LEON3FT_TN0011
+			 ".align 16\n\t"
+#endif
 			 "1:	ldstub	[%1], %0\n\t"
 			 "	cmp	%0, 0\n\t"
 			 "	bne	1b\n\t"
@@ -113,7 +134,16 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 			 : "r" (&_Atomicity_lock<0>::_S_atomicity_lock)
 			 : "memory");
     *__mem += __val;
-    __asm__ __volatile__("stb	%%g0, [%0]"
+    __asm__ __volatile__(
+#ifdef __FIX_LEON3FT_B2BST
+			 "nop\n\t"
+			 "nop\n\t"
+#endif
+			 "stb	%%g0, [%0]\n\t"
+#ifdef __FIX_LEON3FT_B2BST
+			 "nop\n\t"
+			 "nop\n\t"
+#endif
 			 : /* no outputs */
 			 : "r" (&_Atomicity_lock<0>::_S_atomicity_lock)
 			 : "memory");
-- 
2.34.1

